### ISSUE TITLE
addRowData broken in IE8

### ISSUE THREAD

**Issue by dkrey** [free-jqgrid/jqGrid#15]
Hi there,
it seems like adding an empty row doesn't work with ie8 (and possibly others)
This code works fine with FFox and Chrome, but in IE I get
"object doesn't support this property or method":

 var max_rowid = 1+ $(this).jqGrid('getGridParam', 'records') ;
 $(this).jqGrid('addRowData', max_rowid, {}); 

You can also check this example, it results in the same error.
http://www.ok-soft-gmbh.com/jqGrid/OK/navButtons2.htm

Thanks for looking into this :)

**Comment by OlegKi** [comment]
Thank you for the feedback! It's a bug in jqGrid 4.7.0 (see [the line](https://github.com/tonytomov/jqGrid/blob/v4.7.0/js/grid.base.js#L3331)) which I fixed now (see [here](https://github.com/free-jqgrid/jqGrid/commit/935327341ad08d4c3e12da4722aff321f2ce20d8)). Please reload the latest version from [github](https://github.com/free-jqgrid/jqGrid).

I have some technical problems to test the code in old web browsers. I would be appropriate you if you send me more bug reports if you will find a problem. Thank you!

**Comment by dkrey** [comment]
Wow, that was fast, thanks a lot! The error message is gone, although I have to add a trigger('reloadGrid' ) in IE which is not needed for Firefox or Chrome.
Anyway, it's working great now, if I stumble upon anything new I'll let you know :)

**Comment by OlegKi** [comment]
If you could provide more exact description of the problem with `trigger('reloadGrid')` or better a demo, I could reproduce the problem and fix it in the code of jqGrid.

**Comment by dkrey** [comment]
Here's the grid. Without the reloadGrid trigger, Firefox adds the row, the row is rendered and once one leaves the row, the row id is shown. In IE the row count increases, but the new rows aren't rendered until the trigger is called. 

```javascript
$("#order_grid").jqGrid({
		localReader: { repeatitems: false },
		data: emptydata,
		datatype: "clientSide",
		mtype: "GET",
		colNames: ["Pos", "Article No.", "Article", "Vendor","Amount","Size","Price","Order date"],
		colModel: [
				{ name: "id", width: 50, hidden:false, editable:false, sortable: false,sorttype:'int'},
				{ name: "art_no", width: 150, editable:true, sortable: false,sorttype:'text'},
				{ name: "art_name", width: 250, align: "left", editable:true,sortable: false, sorttype:'text'},
				{ name: "vendor", width: 200, align: "left", editable:true,sortable: false, sorttype:'text'},
				{ name: "amount", width: 70, align: "right", editable:true,sorttype:'text' },
				{ name: "size", width: 70, align: "right",editable:true,sorttype:'text'},
				{ name: "prize", width: 70, align: "right",editable:true,sorttype:'text'},
				{
					name: 'order_dat',
					width: 150,
					editable: true,
					edittype:"text",
					formatter: 'date',
					formatoptions: { srcformat:'d.m.Y', newformat:'d.m.Y' },
					editoptions: {
						// dataInit is the client-side event that fires upon initializing the toolbar search field for a column
						// use it to place a third party control to customize the toolbar
						dataInit: function (element) {

							$(element).datepicker({
								id: 'orderDate_datePicker',
								dateFormat: 'dd.mm.yy',
								//minDate: new Date(2010, 0, 1),
								maxDate: new Date(2020, 0, 1),
								showOn: 'focus'

							});
						}
					}
				}
		],
		pager: "#order_pager",
		editurl: 'clientArray',
		rowNum: 30,
		deepempty: true,
		height: 'auto',
		shrinkToFit: false,
		altRows: true,
		altclass: 'gridAltRow',
		rowList: [30, 50, 100],
		sortname: "id",

		onSelectRow: function (rowid) {
			var $self = $(this),
			savedRows = $self.jqGrid("getGridParam", "savedRow");
			if (savedRows.length > 0) {
				$self.jqGrid('saveRow',savedRows[0].id, false);
			}
			current_row_id = rowid;
			$self.jqGrid("editRow", rowid, { keys: true });
		},
		sortorder: "asc",
		viewrecords: true,
		gridview: true,   

		loadError : function(xhr,st,err) {
			$("#order_error_grid").show();
			$("#order_error_grid_content").html("Type: "+st+"; Response: "+err + " "+ "<br/>\n "+
					xhr.responseText+" (HTTP "+xhr.status + " " +xhr.statusText + ")"); 
		},
		// Once the grid is complete, hide some options and add events for editing
		loadComplete: function (data) {
		},              
		autoencode: true,
		caption: "Bestellkopf"
}).navGrid('#order_pager',{view:false, add:false, edit:false,del:false,search:false,refresh:false}, 
		{}, // use default settings for edit
		{}, // use default settings for add
		{},  // delete instead that del:false we need this
		{multipleSearch : false}, // enable the advanced searching
		{closeOnEscape:true} /* allow the view dialog to be closed when user press ESC key*/
		);
	// add first custom button
order_grid.navButtonAdd('#order_pager',
	{
		buttonicon: "ui-icon-plus",
		title: "weitere Zeile",
		caption: "weitere Zeile",
		position: "last",
		onClickButton: function() {
			var $self = $(this);
			var max_rowid = 1+ $self.jqGrid('getGridParam', 'records') ;
			$self.addRowData(max_rowid, {});
			$self.trigger( 'reloadGrid' );
		}
	});
};
```

**Comment by OlegKi** [comment]
Thank you! I do made a bug in the modified line. `inlineNav` just specify `position` parameter of `addRowData` *explicitly*, but you skipped the optional parameter so `addRowData` worked incorrectly. Sorry! I've fixed the code one more time and tested it. Now all seems to work correctly.

I would you strictly recommend to use `undefined` instead of `max_rowid`. jqGrid will generate automatically unique rowid. If you would add Delete row functionality then you can see id duplicates with the code which you `max_rowid`. The usage of `undefined` instead of `addRowData` grantee unique id values.

I would recommend you to make some other changes in the code which I include below
```JavaScript
$("#order_grid").jqGrid({
    data: emptydata,
    colNames: ["Pos", "Article No.", "Article", "Vendor","Amount","Size","Price","Order date"],
    colModel: [
            { name: "id", width: 50, editable: false, sortable: false, sorttype: 'int' },
            { name: "art_no", width: 150, sortable: false },
            { name: "art_name", width: 250, sortable: false },
            { name: "vendor", width: 200,sortable: false },
            { name: "amount", width: 70, align: "right" },
            { name: "size", width: 70, align: "right" },
            { name: "prize", width: 70, align: "right" },
            {
                name: 'order_dat',
                width: 150,
                formatter: 'date',
                formatoptions: { srcformat:'d.m.Y', newformat:'d.m.Y' },
                editoptions: {
                    // dataInit is the client-side event that fires upon initializing the toolbar search field for a column
                    // use it to place a third party control to customize the toolbar
                    dataInit: function (element) {

                        $(element).datepicker({
                            id: 'orderDate_datePicker',
                            dateFormat: 'dd.mm.yy',
                            //minDate: new Date(2010, 0, 1),
                            maxDate: new Date(2020, 0, 1),
                            showOn: 'focus'

                        });
                    }
                }
            }
    ],
    cmTemplate: { editable: true },
    pager: '#order_pager',
    rowNum: 30,
    deepempty: true,
    shrinkToFit: false,
    altRows: true,
    altclass: 'gridAltRow',
    rowList: [30, 50, 100],
    sortname: "id",
    onSelectRow: function (rowid) {
        var $self = $(this),
        savedRows = $self.jqGrid("getGridParam", "savedRow");
        if (savedRows.length > 0) {
            $self.jqGrid('saveRow',savedRows[0].id);
        }
        current_row_id = rowid;
        $self.jqGrid("editRow", rowid);
    },
    sortorder: "asc",
    viewrecords: true,
    loadError : function(xhr,st,err) {
        $("#order_error_grid").show();
        $("#order_error_grid_content").html("Type: "+st+"; Response: "+err + " "+ "<br/>\n "+
                xhr.responseText+" (HTTP "+xhr.status + " " +xhr.statusText + ")");
    },
    autoencode: true,
    caption: "Bestellkopf",
    inlineEditing: {keys: true},
    navOptions: { view:false, add:false, edit:false, del:false, search:false, refresh:false },
    searching: {multipleSearch: true},  // enable the advanced searching
    formViewing: { closeOnEscape: true }
}).jqGrid("navGrid", '#order_pager');
// add first custom button
order_grid.jqGrid("navButtonAdd", '#order_pager',
{
    buttonicon: "ui-icon-plus",
    title: "weitere Zeile",
    caption: "weitere Zeile",
    position: "last",
    onClickButton: function() {
        $(this).jqGrid("addRowData", undefined, {});
    }
});
```
jqGrid set some options (`gridview:true`, `datatype:"local"` and other) automatically based on other options which you specified. Moreover one can specify some options like `inlineEditing: {keys: true}`, `navOptions`, `searching`, `formEditing`, `formDeleting` **as parameter of jqGrid**. In the way you specify default options or callbacks for the corresponding jqGrid methods. The calls of the methods will be shorter and you can have **one place** to specify all options. If you tried to use `inlineNav`, `formatter: "actions"` and direct call of `editRow` in one grid you will good understand the problem. You will have to set `keys: true` option multiple times in relatively deep part of the options.

Best regards
Oleg
P.S. Don't forget to update the sources once more.

**Comment by dkrey** [comment]
Very nice, thanks a lot! This actually taught me a lot (more) about jqgrid.

**Comment by OlegKi** [comment]
You are welcome!

**Comment by dkrey** [comment]
Sorry to bother you again, but I think there's still something wrong with addRowData.
When I use  `$(this).jqGrid("addRowData", undefined, {});` from outside of the grid, I get this error in FireBug:
`
TypeError: R.data.push is not a function
Q,R._index[Q]=R.data.length,R.data.push(T),T={})}!0!==R.altRows||O||("last"===d?...
`

and "object doesn't support this property or method" in IE8

Btw. the row is still added, but the error remains.

This is how I trigger the bug(?)
```javascript
$( "#btnAdd" ).button().click(function( event ) {
     
    event.preventDefault();  
    $("#search_grid").jqGrid("addRowData", undefined, {});
});
```

**Comment by OlegKi** [comment]
It's no problem. I have an interest to fix all existing problems in jqGrid. Do you get the newest code of free jqGrid? I recommend you to refresh the cache too by clicking of F5 or by deleting of local browser cache. Could you try [the demo](http://www.ok-soft-gmbh.com/jqGrid/OK/reloadGrid.htm) which I used for my tests. If you can reproduce some errors on the demo or on youth, please use `jquery.jqGrid.src.js` instead of `jquery.jqGrid.min.js` to localize the error exactly. Please send me always the error description and the row number.

**Comment by dkrey** [comment]
Sure, here's the complete error message from my project:
`TypeError: p.data.push is not a function
http://localhost:8180/hwf_app/lib/jquerygrid-4.7.1/jquery.jqGrid.src.js
Line 4715
`
In your demo I get this:
`TypeError: p is undefined
http://localhost:8180/hwf_app/lib/jquerygrid/jquery.jqGrid.src.js
Line 4619`


This is what I added to the demo (below the grid table tag)
```
    <button id="btnNewPos">Add..</button>
    
        <script>
        $( "#btnNewPos" ).button().click(function( event ) {
            event.preventDefault();  
                var $self = $("#order_grid"), p = $self.jqGrid("getGridParam"), newPos = parseInt(p.records, 10) + 1;
                $(this).jqGrid("addRowData", undefined, {pos: newPos});
        });
    </script>
```

These are the lines from jqgrid
```javascript
			this.each(function() {
				var t = this, p = t.p, datalen = rdata.length;
				ni = p.rownumbers===true ? 1 :0;
```

**Comment by OlegKi** [comment]
The problem is that you use `$(this).jqGrid("addRowData", undefined, {pos: newPos});` inside of ` $( "#btnNewPos" ).button().click(function( event ) {...}`. In the event handler `this` are initialized to `$( "#btnNewPos")`. You need fix the code and use `$self.jqGrid("addRowData", undefined, {pos: newPos});` instead where `$self` variable are defined as `$self = $("#order_grid")` above.

**Comment by dkrey** [comment]
Oh my bad, sorry about this. 
The problem I had with my project was that I also added `data: {}` as a grid parameter. 
This caused `TypeError: p.data.push is not a function`.
Thanks again for your time.

**Comment by OlegKi** [comment]
`data` must be array. So empty `data` should be `data: []`.
